### PR TITLE
Stats: force new user to purchase page (still behind feature flag)

### PR DESCRIPTION
**Pull request by grzegorz-cp**
Related to #86130 

## Proposed Changes

* Adding a check for the blog creation date to enable new purchase flow for Jetpack sites

## Testing Instructions

* create a JN blog and connect Jetpack
* open the live branch and navigate to the Stats page for the new blog
* the Stats page should load
* apply `stats/checkout-flows-v2` feature flag and verify that you are redirected to a purchase page

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/stats/site.jsx [hunk 1]
```diff
@@ -613,6 +623,15 @@ export default connect(
 		const upsellModalView =
 			config.isEnabled( 'stats/paid-wpcom-v2' ) && getUpsellModalView( state, siteId );
 
+		const siteCreatedTimeStamp = getSiteOptions( state, siteId ?? 0 )?.created_at;
+
+		// TODO: update the date to the release date when the feature is ready.
+		const redirectToPurchase =
+			config.isEnabled( 'stats/checkout-flows-v2' ) &&
+			isJetpack &&
```
Comment (kangzj) [inline client/my-sites/stats/site.jsx]: I think we'll need to exclude atomic sites here, which is Jetpack and wpcom

Comment (kangzj) [reply client/my-sites/stats/site.jsx]: example: https://github.com/Automattic/wp-calypso/blob/7436c47785d5ba44281bfd92c9952581a205517e/client/my-sites/stats/stats-notices/index.tsx#L75

File: client/my-sites/stats/site.jsx [hunk 2]
```diff
@@ -613,6 +623,15 @@ export default connect(
...
+			siteCreatedTimeStamp &&
+			new Date( siteCreatedTimeStamp ) > new Date( '2024-01-01' );
```
Comment (dognose24) [inline client/my-sites/stats/site.jsx:638]: Is this compared date `2024-01-01` a temporary one?

File: client/my-sites/stats/site.jsx [hunk 3]
```diff
@@ -540,6 +544,12 @@ class StatsSite extends Component {
 		} else if ( this.props.showEnableStatsModule ) {
 			return this.renderEnableStatsModule();
 		}
+
+		// render purchase flow for Jetpack sites created after February 2024
+		if ( this.props.redirectToPurchase && this.props.slug ) {
+			page.redirect( `/stats/purchase/${ this.props.slug }` );
```
Comment (kangzj) [inline client/my-sites/stats/site.jsx:550]: Should we `return` after the redirection here?

File: client/my-sites/stats/site.jsx (hunk 2 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/site.jsx:638]: Yes, we will change it to the release date just before the release. I used this date so people can test it on the live branches and don't have to checkout branches for new pull requests.

File: client/my-sites/stats/site.jsx (hunk 3 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/site.jsx:550]: I don't think it's needed, but I can add it for good measure.

**Review by kangzj — APPROVED** [review]
LGTM 👍

File: client/my-sites/stats/site.jsx [hunk 4]
```diff
@@ -613,6 +628,16 @@ export default connect(
 		const upsellModalView =
 			config.isEnabled( 'stats/paid-wpcom-v2' ) && getUpsellModalView( state, siteId );
 
+		const siteCreatedTimeStamp = getSiteOptions( state, siteId ?? 0 )?.created_at;
+
+		// TODO: update the date to the release date when the feature is ready.
+		const redirectToPurchase =
+			config.isEnabled( 'stats/checkout-flows-v2' ) &&
+			isJetpack &&
```
Comment (kangzj) [inline client/my-sites/stats/site.jsx]:
```suggestion
```

The condition wouldn't be necessary anymore after using `isSiteJetpackNotAtomic` 🙂

File: client/my-sites/stats/site.jsx [hunk 5]
```diff
@@ -613,6 +628,15 @@ export default connect(
 		const upsellModalView =
 			config.isEnabled( 'stats/paid-wpcom-v2' ) && getUpsellModalView( state, siteId );
 
+		const siteCreatedTimeStamp = getSiteOptions( state, siteId ?? 0 )?.created_at;
+
+		// TODO: update the date to the release date when the feature is ready.
+		const redirectToPurchase =
+			config.isEnabled( 'stats/checkout-flows-v2' ) &&
+			isSiteJetpackNotAtomic &&
+			siteCreatedTimeStamp &&
```
Comment (kangzj) [inline client/my-sites/stats/site.jsx:637]: Hi @grzegorz-cp,

Me and @a8ck3n were talking about this today, and we suspected that the created time might not be the time we are after here. We might be after the 'connected' time, which is the time when the site is connected 🤔 

I'm happy for it to merge and iterate still tho 😄

Comment (kangzj) [reply client/my-sites/stats/site.jsx:637]: Sorry for pointing probably the wrong data for you.

Comment (grzegorz-cp) [reply client/my-sites/stats/site.jsx:637]: Let's merge it and potentially change it in the next PRs. I'm curious about the scenarios where `connected` makes a difference.

**Pull request merged**